### PR TITLE
enhance: speed up array-equal operator via inverted index

### LINKED ISSUE #33632: [Enhancement]: speed up array-equal operator via inverted index
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What would you like to be added?

Array-equal operator can be speed up via inverted index. We can get the candidates which contain all the elements of the array, and then use the raw data to do the post-filtering.

### Why is this needed?

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by longjiquan**
fix: #33632

### CODE REVIEW

File: internal/core/src/index/ScalarIndex.h [hunk 1]
```diff
@@ -50,6 +50,27 @@ class ScalarIndex : public IndexBase {
     virtual const TargetBitmap
     In(size_t n, const T* values) = 0;
 
+    virtual const TargetBitmap
+    InV2(size_t n,
```
Comment (zhagnlu) [inline internal/core/src/index/ScalarIndex.h:54]: rename InApplyFilter or others, not use v1 v2 etc.

File: internal/core/src/index/ScalarIndex.h [hunk 2]
```diff
@@ -50,6 +50,27 @@ class ScalarIndex : public IndexBase {
...
+         const T* values,
+         const std::function<bool(size_t /* offset */)>& filter) {
+        auto res = In(n, values);
+        for (size_t offset = 0; offset < res.size(); offset++) {
+            res[offset] = res[offset] && filter(offset);
```
Comment (zhagnlu) [inline internal/core/src/index/ScalarIndex.h:59]: res[offset]  &&= filter(offset);

File: internal/core/src/index/ScalarIndex.h [hunk 3]
```diff
@@ -50,6 +50,27 @@ class ScalarIndex : public IndexBase {
...
+    virtual void
+    InV3(size_t n,
+         const T* values,
+         const std::function<void(size_t /* offset */)>& callback) {
+        auto res = In(n, values);
+        for (size_t offset = 0; offset < res.size(); offset++) {
+            callback(offset);
+        }
+    }
+
```
Comment (zhagnlu) [inline internal/core/src/index/ScalarIndex.h:73]: anyway, is that necessary add this interface in base class, other subclass need this function ?

File: internal/core/src/exec/expression/Expr.h [hunk 4]
```diff
@@ -280,6 +280,22 @@ class SegmentExpr : public Expr {
         return result;
     }
 
+    template <typename T, typename FUNC, typename... ValTypes>
+    void
+    ProcessIndexChunksV2(FUNC func, ValTypes... values) {
```
Comment (zhagnlu) [inline internal/core/src/exec/expression/Expr.h:285]: not use v2 name,  why not use upper ProcessIndexChunk mode?

Comment (longjiquan) [reply internal/core/src/exec/expression/Expr.h:285]: ProcessIndexChunk will move the cursor, but I only need to iterate all index chunks.

File: internal/core/src/index/ScalarIndex.h (hunk 3 above)
Comment (longjiquan) [reply internal/core/src/index/ScalarIndex.h:73]: Yes, you are right that we cannot push down the filter or callback to all indexes. But I don't know how to design the interface, should we check whether the index is InvertedIndex?

File: internal/core/src/index/ScalarIndex.h [hunk 5]
```diff
@@ -50,6 +50,27 @@ class ScalarIndex : public IndexBase {
...
+    virtual void
+    InV3(size_t n,
+         const T* values,
+         const std::function<void(size_t /* offset */)>& callback) {
+        auto res = In(n, values);
+        for (size_t offset = 0; offset < res.size(); offset++) {
+            callback(offset);
+        }
```
[the remark this answers is not in the document]
Comment (longjiquan) [reply internal/core/src/index/ScalarIndex.h:71]: Good catch! The callback should be called when res[offset] is true and I'll fix this.

File: internal/core/src/index/ScalarIndex.h (hunk 2 above)
Comment (longjiquan) [reply internal/core/src/index/ScalarIndex.h:59]: OK.

File: internal/core/src/index/ScalarIndex.h (hunk 1 above)
Comment (longjiquan) [reply internal/core/src/index/ScalarIndex.h:54]: OK.

**Comment by longjiquan** [comment]
comment fixed in https://github.com/milvus-io/milvus/pull/33633/commits/2064c8f6985e2cdf7ac4d6faa9aa92cd88738581, cc @zhagnlu 
and ut added in https://github.com/milvus-io/milvus/pull/33633/commits/b9f8647f626cd8f2fc2274ca39d2bbeeadbd1283

**Comment by longjiquan** [comment]
rerun ut

File: internal/core/src/exec/expression/UnaryExpr.cpp [hunk 6]
```diff
@@ -196,6 +262,106 @@ PhyUnaryRangeFilterExpr::ExecRangeVisitorImplArray() {
...
+        ProcessIndexChunks<IndexInnerType>([this, &val, reverse](Index* _) {
+            std::unordered_set<IndexInnerType> elements;
+            boost::container::vector<IndexInnerType> elems;
+            for (auto const& element : val.array()) {
+                auto e = GetValueFromProto<IndexInnerType>(element);
+                if (elements.find(e) == elements.end()) {
+                    elems.push_back(e);
+                }
+                elements.insert(e);
+            }
```
Comment (zhagnlu) [inline internal/core/src/exec/expression/UnaryExpr.cpp:295]: no need to use element set, just std::find(elems.begin(), elems.end(), target) ! = elems.end()

File: internal/core/src/exec/expression/UnaryExpr.cpp [hunk 7]
```diff
@@ -196,6 +262,106 @@ PhyUnaryRangeFilterExpr::ExecRangeVisitorImplArray() {
...
+
+            // filtering by index, get candidates.
+            auto size_per_chunk = segment_->size_per_chunk();
+            auto retrieve = [ size_per_chunk, this ](int64_t offset) -> auto {
+                auto chunk_idx = offset / size_per_chunk;
+                auto chunk_offset = offset % size_per_chunk;
+                const auto& chunk =
+                    segment_->template chunk_data<milvus::ArrayView>(field_id_,
```
Comment (zhagnlu) [inline internal/core/src/exec/expression/UnaryExpr.cpp:303]: this interface only for sealed segment, what about growing ?

File: internal/core/src/exec/expression/UnaryExpr.cpp (hunk 6 above)
Comment (longjiquan) [reply internal/core/src/exec/expression/UnaryExpr.cpp:295]: OK.

**Pull request merged**